File: cmd/cshared/internal/capi/cipher.go
package capi

import (
	"github.com/everanium/itb"
	"github.com/everanium/itb/hashes"
)

// Encrypt encrypts plaintext using the (noise, data, start) seed
// trio. All three handles must wrap seeds of the same native hash
// width; otherwise StatusSeedWidthMix is returned.
//
// out is caller-allocated. On success the encrypted payload is
// written to out[:n] and n is reported back through outLen.
// If cap(out) is too small, StatusBufferTooSmall is returned and
// outLen is set to the size that would have been written so the
// caller can resize and retry.
func Encrypt(noise, data, start HandleID, plaintext, out []byte) (n int, st Status) {
	defer recoverPanic(&st, StatusEncryptFailed)

	ns, ds, ss, st := resolveTriple(noise, data, start)
	if st != StatusOK {
		return 0, st
	}
	enc, err := encryptDispatch(ns, ds, ss, plaintext)
	if err != nil {
		setLastErr(StatusEncryptFailed)
		return 0, StatusEncryptFailed
	}
	if len(enc) > len(out) {
		setLastErr(StatusBufferTooSmall)
		return len(enc), StatusBufferTooSmall
	}
	copy(out, enc)
	return len(enc), StatusOK
}

// Decrypt is the inverse of Encrypt. Same caller-allocated-buffer
// convention; outLen receives the recovered plaintext length on
// success or the required capacity on StatusBufferTooSmall.
func Decrypt(noise, data, start HandleID, ciphertext, out []byte) (n int, st Status) {
	defer recoverPanic(&st, StatusDecryptFailed)

	ns, ds, ss, st := resolveTriple(noise, data, start)
	if st != StatusOK {
		return 0, st
	}
	plain, err := decryptDispatch(ns, ds, ss, ciphertext)
	if err != nil {
		setLastErr(StatusDecryptFailed)
		return 0, StatusDecryptFailed
	}
	if len(plain) > len(out) {
		setLastErr(StatusBufferTooSmall)
		return len(plain), StatusBufferTooSmall
	}
	copy(out, plain)
	return len(plain), StatusOK
}

func resolveTriple(noise, data, start HandleID) (ns, ds, ss *SeedHandle, st Status) {
	ns, st = resolve(noise)
	if st != StatusOK {
		return nil, nil, nil, st
	}
	ds, st = resolve(data)
	if st != StatusOK {
		return nil, nil, nil, st
	}
	ss, st = resolve(start)
	if st != StatusOK {
		return nil, nil, nil, st
	}
	if ns.width != ds.width || ds.width != ss.width {
		setLastErr(StatusSeedWidthMix)
		return nil, nil, nil, StatusSeedWidthMix
	}
	return ns, ds, ss, StatusOK
}

func encryptDispatch(ns, ds, ss *SeedHandle, plaintext []byte) ([]byte, error) {
	switch ns.width {
	case hashes.W128:
		return itb.Encrypt128(ns.seed128, ds.seed128, ss.seed128, plaintext)
	case hashes.W256:
		return itb.Encrypt256(ns.seed256, ds.seed256, ss.seed256, plaintext)
	case hashes.W512:
		return itb.Encrypt512(ns.seed512, ds.seed512, ss.seed512, plaintext)
	}
	return nil, errInternal
}

func decryptDispatch(ns, ds, ss *SeedHandle, ciphertext []byte) ([]byte, error) {
	switch ns.width {
	case hashes.W128:
		return itb.Decrypt128(ns.seed128, ds.seed128, ss.seed128, ciphertext)
	case hashes.W256:
		return itb.Decrypt256(ns.seed256, ds.seed256, ss.seed256, ciphertext)
	case hashes.W512:
		return itb.Decrypt512(ns.seed512, ds.seed512, ss.seed512, ciphertext)
	}
	return nil, errInternal
}

// errInternal is the sentinel returned by dispatch helpers when the
// cached width on a handle does not match any known case (should be
// unreachable; kept as a defensive placeholder).
var errInternal = capiError("capi: internal width mismatch")

type capiError string

func (e capiError) Error() string { return string(e) }

// Encrypt3 is the Triple Ouroboros counterpart of Encrypt. It takes
// 7 seed handles (1 shared noise + 3 data + 3 start) and produces
// one ciphertext that splits the plaintext into three interleaved
// snake payloads. The on-wire format is the same shape as the
// non-Triple ciphertext (same nonce + dimensions header + raw
// container) — only the internal split / interleave differs.
//
// All 7 handles must share the same native hash width and must be
// pairwise distinct (the underlying itb.Encrypt3x* enforces the
// seven-seed isolation invariant). Same caller-allocated-buffer
// convention as Encrypt: returned n carries the bytes written on
// success or the required capacity on StatusBufferTooSmall.
func Encrypt3(
	noise, data1, data2, data3, start1, start2, start3 HandleID,
	plaintext, out []byte,
) (n int, st Status) {
	defer recoverPanic(&st, StatusEncryptFailed)

	ns, ds1, ds2, ds3, ss1, ss2, ss3, st := resolveSeven(
		noise, data1, data2, data3, start1, start2, start3,
	)
	if st != StatusOK {
		return 0, st
	}
	enc, err := encryptTripleDispatch(ns, ds1, ds2, ds3, ss1, ss2, ss3, plaintext)
	if err != nil {
		setLastErr(StatusEncryptFailed)
		return 0, StatusEncryptFailed
	}
	if len(enc) > len(out) {
		setLastErr(StatusBufferTooSmall)
		return len(enc), StatusBufferTooSmall
	}
	copy(out, enc)
	return len(enc), StatusOK
}

// Decrypt3 is the inverse of Encrypt3. Same convention.
func Decrypt3(
	noise, data1, data2, data3, start1, start2, start3 HandleID,
	ciphertext, out []byte,
) (n int, st Status) {
	defer recoverPanic(&st, StatusDecryptFailed)

	ns, ds1, ds2, ds3, ss1, ss2, ss3, st := resolveSeven(
		noise, data1, data2, data3, start1, start2, start3,
	)
	if st != StatusOK {
		return 0, st
	}
	plain, err := decryptTripleDispatch(ns, ds1, ds2, ds3, ss1, ss2, ss3, ciphertext)
	if err != nil {
		setLastErr(StatusDecryptFailed)
		return 0, StatusDecryptFailed
	}
	if len(plain) > len(out) {
		setLastErr(StatusBufferTooSmall)
		return len(plain), StatusBufferTooSmall
	}
	copy(out, plain)
	return len(plain), StatusOK
}

func resolveSeven(
	noise, data1, data2, data3, start1, start2, start3 HandleID,
) (ns, ds1, ds2, ds3, ss1, ss2, ss3 *SeedHandle, st Status) {
	ids := [7]HandleID{noise, data1, data2, data3, start1, start2, start3}
	resolved := [7]*SeedHandle{}
	for i, id := range ids {
		h, s := resolve(id)
		if s != StatusOK {
			return nil, nil, nil, nil, nil, nil, nil, s
		}
		resolved[i] = h
	}
	w := resolved[0].width
	for _, h := range resolved[1:] {
		if h.width != w {
			setLastErr(StatusSeedWidthMix)
			return nil, nil, nil, nil, nil, nil, nil, StatusSeedWidthMix
		}
	}
	return resolved[0], resolved[1], resolved[2], resolved[3],
		resolved[4], resolved[5], resolved[6], StatusOK
}

func encryptTripleDispatch(
	ns, ds1, ds2, ds3, ss1, ss2, ss3 *SeedHandle, plaintext []byte,
) ([]byte, error) {
	switch ns.width {
	case hashes.W128:
		return itb.Encrypt3x128(
			ns.seed128, ds1.seed128, ds2.seed128, ds3.seed128,
			ss1.seed128, ss2.seed128, ss3.seed128, plaintext)
	case hashes.W256:
		return itb.Encrypt3x256(
			ns.seed256, ds1.seed256, ds2.seed256, ds3.seed256,
			ss1.seed256, ss2.seed256, ss3.seed256, plaintext)
	case hashes.W512:
		return itb.Encrypt3x512(
			ns.seed512, ds1.seed512, ds2.seed512, ds3.seed512,
			ss1.seed512, ss2.seed512, ss3.seed512, plaintext)
	}
	return nil, errInternal
}

func decryptTripleDispatch(
	ns, ds1, ds2, ds3, ss1, ss2, ss3 *SeedHandle, ciphertext []byte,
) ([]byte, error) {
	switch ns.width {
	case hashes.W128:
		return itb.Decrypt3x128(
			ns.seed128, ds1.seed128, ds2.seed128, ds3.seed128,
			ss1.seed128, ss2.seed128, ss3.seed128, ciphertext)
	case hashes.W256:
		return itb.Decrypt3x256(
			ns.seed256, ds1.seed256, ds2.seed256, ds3.seed256,
			ss1.seed256, ss2.seed256, ss3.seed256, ciphertext)
	case hashes.W512:
		return itb.Decrypt3x512(
			ns.seed512, ds1.seed512, ds2.seed512, ds3.seed512,
			ss1.seed512, ss2.seed512, ss3.seed512, ciphertext)
	}
	return nil, errInternal
}

File: bindings/nodejs/src/cipher.ts
// Low-level encrypt / decrypt entry points (Single + Triple, plain +
// authenticated). Mirrors bindings/python/itb/_ffi.py top-level
// `encrypt`, `decrypt`, `encrypt_triple`, `decrypt_triple`,
// `encrypt_auth`, `decrypt_auth`, `encrypt_auth_triple`,
// `decrypt_auth_triple`.
//
// Output sizing follows the libitb probe-then-write protocol: the
// first call with `cap=0` returns `Status.BufferTooSmall` and writes
// the required size into `outLen`; the second call with the
// allocated buffer fills it.
//
// All seeds passed to one call must share the same native hash
// width. Mixing widths raises `ITBError(SeedWidthMix)`.

import { errorFromStatus } from './errors.js';
import type { MAC } from './mac.js';
import {
  ITB_Decrypt,
  ITB_Decrypt3,
  ITB_DecryptAuth,
  ITB_DecryptAuth3,
  ITB_Encrypt,
  ITB_Encrypt3,
  ITB_EncryptAuth,
  ITB_EncryptAuth3,
} from './native.js';
import type { Seed } from './seed.js';
import { Status } from './status.js';

type Handle = bigint | number;

type SingleFn = (
  noise: Handle,
  data: Handle,
  start: Handle,
  payload: Uint8Array,
  ptlen: number,
  out: Uint8Array | null,
  outCap: number,
  outLen: [number | bigint],
) => number;

type TripleFn = (
  noise: Handle,
  data1: Handle,
  data2: Handle,
  data3: Handle,
  start1: Handle,
  start2: Handle,
  start3: Handle,
  payload: Uint8Array,
  ptlen: number,
  out: Uint8Array | null,
  outCap: number,
  outLen: [number | bigint],
) => number;

type AuthSingleFn = (
  noise: Handle,
  data: Handle,
  start: Handle,
  mac: Handle,
  payload: Uint8Array,
  ptlen: number,
  out: Uint8Array | null,
  outCap: number,
  outLen: [number | bigint],
) => number;

type AuthTripleFn = (
  noise: Handle,
  data1: Handle,
  data2: Handle,
  data3: Handle,
  start1: Handle,
  start2: Handle,
  start3: Handle,
  mac: Handle,
  payload: Uint8Array,
  ptlen: number,
  out: Uint8Array | null,
  outCap: number,
  outLen: [number | bigint],
) => number;

function ensureBytes(payload: Uint8Array, label: string): void {
  if (!(payload instanceof Uint8Array)) {
    throw new TypeError(`${label} must be a Uint8Array`);
  }
}

function runSingle(
  fn: SingleFn,
  noise: Seed,
  data: Seed,
  start: Seed,
  payload: Uint8Array,
): Uint8Array {
  const probe: [number | bigint] = [0];
  let rc = fn(noise.handle, data.handle, start.handle, payload, payload.length, null, 0, probe);
  if (rc === Status.Ok) {
    return new Uint8Array(0);
  }
  if (rc !== Status.BufferTooSmall) {
    throw errorFromStatus(rc);
  }
  const need = Number(probe[0]);
  const out = new Uint8Array(need);
  const filled: [number | bigint] = [0];
  rc = fn(noise.handle, data.handle, start.handle, payload, payload.length, out, need, filled);
  if (rc !== Status.Ok) {
    throw errorFromStatus(rc);
  }
  return out.subarray(0, Number(filled[0]));
}

function runTriple(
  fn: TripleFn,
  noise: Seed,
  data1: Seed,
  data2: Seed,
  data3: Seed,
  start1: Seed,
  start2: Seed,
  start3: Seed,
  payload: Uint8Array,
): Uint8Array {
  const probe: [number | bigint] = [0];
  let rc = fn(
    noise.handle,
    data1.handle, data2.handle, data3.handle,
    start1.handle, start2.handle, start3.handle,
    payload, payload.length,
    null, 0, probe,
  );
  if (rc === Status.Ok) {
    return new Uint8Array(0);
  }
  if (rc !== Status.BufferTooSmall) {
    throw errorFromStatus(rc);
  }
  const need = Number(probe[0]);
  const out = new Uint8Array(need);
  const filled: [number | bigint] = [0];
  rc = fn(
    noise.handle,
    data1.handle, data2.handle, data3.handle,
    start1.handle, start2.handle, start3.handle,
    payload, payload.length,
    out, need, filled,
  );
  if (rc !== Status.Ok) {
    throw errorFromStatus(rc);
  }
  return out.subarray(0, Number(filled[0]));
}

function runAuthSingle(
  fn: AuthSingleFn,
  noise: Seed,
  data: Seed,
  start: Seed,
  mac: MAC,
  payload: Uint8Array,
): Uint8Array {
  const probe: [number | bigint] = [0];
  let rc = fn(
    noise.handle, data.handle, start.handle,
    mac.handle,
    payload, payload.length,
    null, 0, probe,
  );
  if (rc === Status.Ok) {
    return new Uint8Array(0);
  }
  if (rc !== Status.BufferTooSmall) {
    throw errorFromStatus(rc);
  }
  const need = Number(probe[0]);
  const out = new Uint8Array(need);
  const filled: [number | bigint] = [0];
  rc = fn(
    noise.handle, data.handle, start.handle,
    mac.handle,
    payload, payload.length,
    out, need, filled,
  );
  if (rc !== Status.Ok) {
    throw errorFromStatus(rc);
  }
  return out.subarray(0, Number(filled[0]));
}

function runAuthTriple(
  fn: AuthTripleFn,
  noise: Seed,
  data1: Seed,
  data2: Seed,
  data3: Seed,
  start1: Seed,
  start2: Seed,
  start3: Seed,
  mac: MAC,
  payload: Uint8Array,
): Uint8Array {
  const probe: [number | bigint] = [0];
  let rc = fn(
    noise.handle,
    data1.handle, data2.handle, data3.handle,
    start1.handle, start2.handle, start3.handle,
    mac.handle,
    payload, payload.length,
    null, 0, probe,
  );
  if (rc === Status.Ok) {
    return new Uint8Array(0);
  }
  if (rc !== Status.BufferTooSmall) {
    throw errorFromStatus(rc);
  }
  const need = Number(probe[0]);
  const out = new Uint8Array(need);
  const filled: [number | bigint] = [0];
  rc = fn(
    noise.handle,
    data1.handle, data2.handle, data3.handle,
    start1.handle, start2.handle, start3.handle,
    mac.handle,
    payload, payload.length,
    out, need, filled,
  );
  if (rc !== Status.Ok) {
    throw errorFromStatus(rc);
  }
  return out.subarray(0, Number(filled[0]));
}

/** Encrypts plaintext under the (noise, data, start) seed trio. */
export function encrypt(
  noise: Seed,
  data: Seed,
  start: Seed,
  plaintext: Uint8Array,
): Uint8Array {
  ensureBytes(plaintext, 'plaintext');
  return runSingle(ITB_Encrypt, noise, data, start, plaintext);
}

/** Decrypts ciphertext produced by `encrypt` under the same seed trio. */
export function decrypt(
  noise: Seed,
  data: Seed,
  start: Seed,
  ciphertext: Uint8Array,
): Uint8Array {
  ensureBytes(ciphertext, 'ciphertext');
  return runSingle(ITB_Decrypt, noise, data, start, ciphertext);
}

/**
 * Triple-Ouroboros encrypt over seven seeds.
 *
 * Splits plaintext across three interleaved snake payloads. The
 * on-wire ciphertext format is the same shape as `encrypt` — only
 * the internal split / interleave differs. All seven seeds must
 * share the same native hash width and be pairwise distinct
 * handles.
 */
export function encryptTriple(
  noise: Seed,
  data1: Seed, data2: Seed, data3: Seed,
  start1: Seed, start2: Seed, start3: Seed,
  plaintext: Uint8Array,
): Uint8Array {
  ensureBytes(plaintext, 'plaintext');
  return runTriple(ITB_Encrypt3, noise, data1, data2, data3, start1, start2, start3, plaintext);
}

/** Inverse of `encryptTriple`. */
export function decryptTriple(
  noise: Seed,
  data1: Seed, data2: Seed, data3: Seed,
  start1: Seed, start2: Seed, start3: Seed,
  ciphertext: Uint8Array,
): Uint8Array {
  ensureBytes(ciphertext, 'ciphertext');
  return runTriple(ITB_Decrypt3, noise, data1, data2, data3, start1, start2, start3, ciphertext);
}

/** Authenticated single-Ouroboros encrypt with MAC-Inside-Encrypt. */
export function encryptAuth(
  noise: Seed,
  data: Seed,
  start: Seed,
  mac: MAC,
  plaintext: Uint8Array,
): Uint8Array {
  ensureBytes(plaintext, 'plaintext');
  return runAuthSingle(ITB_EncryptAuth, noise, data, start, mac, plaintext);
}

/**
 * Authenticated single-Ouroboros decrypt. Raises `ITBError` with
 * code `Status.MacFailure` on tampered ciphertext / wrong MAC key.
 */
export function decryptAuth(
  noise: Seed,
  data: Seed,
  start: Seed,
  mac: MAC,
  ciphertext: Uint8Array,
): Uint8Array {
  ensureBytes(ciphertext, 'ciphertext');
  return runAuthSingle(ITB_DecryptAuth, noise, data, start, mac, ciphertext);
}

/** Authenticated Triple-Ouroboros encrypt (7 seeds + MAC). */
export function encryptAuthTriple(
  noise: Seed,
  data1: Seed, data2: Seed, data3: Seed,
  start1: Seed, start2: Seed, start3: Seed,
  mac: MAC,
  plaintext: Uint8Array,
): Uint8Array {
  ensureBytes(plaintext, 'plaintext');
  return runAuthTriple(
    ITB_EncryptAuth3,
    noise, data1, data2, data3, start1, start2, start3, mac, plaintext,
  );
}

/** Authenticated Triple-Ouroboros decrypt. */
export function decryptAuthTriple(
  noise: Seed,
  data1: Seed, data2: Seed, data3: Seed,
  start1: Seed, start2: Seed, start3: Seed,
  mac: MAC,
  ciphertext: Uint8Array,
): Uint8Array {
  ensureBytes(ciphertext, 'ciphertext');
  return runAuthTriple(
    ITB_DecryptAuth3,
    noise, data1, data2, data3, start1, start2, start3, mac, ciphertext,
  );
}
